graylog2-server/src/main/java/org/graylog2/indexer/messages/SerializationMemoizingMessage.java

```diff
@@ -0,0 +1,165 @@
+                              Meter invalidTimeStampMeter) {}
+
+    public SerializationMemoizingMessage(Message delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public synchronized byte[] serialize(SerializationContext context)
+            throws JsonProcessingException {
+
+        final var objectMapper = context.objectMapper();
+        final var invalidTimestampMeter = context.invalidTimestampMeter();
+
+        CacheEntry cachedEntry = null;
+        if (lastSerializationResult != null) {
+            cachedEntry = lastSerializationResult.get();
+            if (cachedEntry == null) {
+                RATE_LIMITED_LOG.warn("The JVM cleared a cached serialized message because of memory pressure. " +
+                        "This has a performance impact. Please adjust the memory configuration to assign more " +
+                        "heap memory to the JVM.");
+            }
+        }
+
+        if (cachedEntry == null || !cachedEntry.objectMapper().equals(objectMapper)) {
+            final var tsMeter = new Meter();
+            final var serializedBytes = delegate.serialize(new DefaultSerializationContext(objectMapper, tsMeter));
+            cachedEntry = new CacheEntry(serializedBytes, objectMapper, tsMeter);
+            this.lastSerializationResult = new SoftReference<>(cachedEntry);
+        }
+
+        invalidTimestampMeter.mark(cachedEntry.invalidTimeStampMeter().getCount());
+        return cachedEntry.serializedBytes();
+    }
+
+    // only straight-forward delegations below this line
+
+    @Override
+    public int hashCode() {
+        return delegate.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof SerializationMemoizingMessage serializationMemoizingMessage)) {
+            return false;
+        }
+        return delegate.equals(serializationMemoizingMessage.delegate);
+    }
+
+    @Override
+    @Deprecated
+    public String getId() {
+        return delegate.getId();
+    }
+
+    @Override
+    public String getMessageId() {
+        return delegate.getMessageId();
+    }
+
+    @Override
+    public long getSize() {
+        return delegate.getSize();
+    }
+
+    @Override
+    public DateTime getReceiveTime() {
+        return delegate.getReceiveTime();
+    }
+
+    @Override
+    public Map<String, Object> toElasticSearchObject(ObjectMapper objectMapper, @Nonnull Meter invalidTimestampMeter) {
+        return delegate.toElasticSearchObject(objectMapper, invalidTimestampMeter);
+    }
+
+    @Override
+    public DateTime getTimestamp() {
+        return delegate.getTimestamp();
+    }
+
+    @Override
+    public boolean supportsFailureHandling() {
+        return delegate.supportsFailureHandling();
+    }
+
+    @Override
+    public Set<IndexSet> getIndexSets() {
+        return delegate.getIndexSets();
+    }
+
+    @Override
+    public Map<String, Object> getFields() {
+        return delegate.getFields();
+    }
```

Comment:
Shouldn't we make immutable copies here because this class implements `ImmutableMessage`?

Reply:
The delegate already returns an immutable copy.

Reply:
Ah, true.